src/blocking/i2s.rs

```diff
@@ -0,0 +1,39 @@
+//! Blocking I2S API
+
+/// Blocking read
+pub trait Read<W> {
+    /// Error type
+    type Error;
+
+    /// Reads enough bytes from the slave to fill `left_words` and `right_words`.
+    fn try_read<'w>(
+        &mut self,
+        left_words: &'w mut [W],
+        right_words: &'w mut [W],
+    ) -> Result<(), Self::Error>;
+}
+
+/// Blocking write
+pub trait Write<W> {
+    /// Error type
+    type Error;
+
+    /// Sends `left_words` and `right_words` to the slave.
+    fn try_write<'w>(
+        &mut self,
+        left_words: &'w [W],
+        right_words: &'w [W],
+    ) -> Result<(), Self::Error>;
+}
+
+/// Blocking write (iterator version)
+pub trait WriteIter<W> {
+    /// Error type
+    type Error;
+
+    /// Sends `left_words` and `right_words` to the slave.
+    fn try_write<LW, RW>(&mut self, left_words: LW, right_words: RW) -> Result<(), Self::Error>
```

Comment:
Do we need `LW` and `RW` separate for some reason?

Reply:
For PCM mode it’s not needed at all, for the other modes it’s not technically needed but could be semantically useful. For the other modes first left then right is written to the data register, or for Philips mode it’s synced to the WS clock. But I’m very new to this and have little practical knowledge. I guess the final decision depends on what abstraction level is the right one for the embedded-hal lib. The more high level left/right channel semantics (or even multichannel) can definitely be left to individual drivers to supply IMO.

Reply:
Separate `LW` and `RW` would allow different types for the buffers as long as they can be converted to a `W` iterator. I would rather not require them to be the same type if it is not necessary.

Reply:
Just out of curiosity, when would it make sense to have different types for the left and right channel? To my understanding they would always be interleaved when written and an application would most likely write both a left and right word at the same time.

Reply:
> Separate LW and RW would allow different types for the buffers as long as they can be converted to a W iterator.

Is there a situation in which this is useful? from my _extremely limited_ i2s experience the left and right channels are always interleaved and usually going to be run at the same sampling rate / bit depth?

Afaik we usually have separate methods for iterator based approaches, rather than trying to be generic over both. For example, something like `fn try_write<W>(&mut self, left_words: &[W], right_words: &[W]) -> Result<(), Self::Error>` vs. `fn try_write_iter<W, I: Iterator<W>>(&mut self, left_words: I, right_words: I)`

Reply:
I would prefer a single interleaved buffer and would be perfectly happy with having a higher level abstraction (with separate left/right etc) in the driver crate instead. That way stereo and multichannel DACs can be supported with better semantics.

Reply:
I thought separate types would improve support for 1-channel "mono" communication. e.g. maybe one wants to have a struct that generates numbers on the fly for the left channel and just loop over a zero-filled buffer for the right channel.

However, I just had a look at the multichannel description (TDM, although it is not standard) and it seems anything higher level than a raw buffer would not be a good fit.
I will update this PR soon.

Reply:
Sounds good. The new interface looks much cleaner BTW.
I'll try to play with an implementation of it in the stm32f4xx-hal crate I get some time, mostly for me to learn the ecosystem better.

Reply:
Now when I have some more knowledge with I2S the separate left and right words might not be such a bad idea after all. What supports this is the fact that a full frame is considered transmitted only after both left and right has been sent. The word size can be 16, 24 or 32 bits. In the STM32F4 there is a special status flag (CHSIDE) which can be used together with TXE to determine which word to send next. It would be good to be able to take advantage of that.

I have also read the data sheet for the CS4385 which is a 8 channel DAC. In TDM mode it can receive all 8 channels over a SAI interface, which is basically a more modular version of I2S. In that case it can send 256 bits after only one WS trigger, but that behavior is not possible with standard I2S. My conclusion is that it’s perfectly ok for the HAL to focus on providing a 2 channel abstraction, which is all that standard I2S will be able to do.

Could you revert back the last change @eldruin? Sorry for the extra rounds 🙂

Reply:
@maxekman Very interesting :) Sure, no problem. I just reverted the last change in the i2s branch. The interleaved version can be found in the `i2s-interleaved` branch.

Reply:
Thanks @eldruin, will give it a go later today.